spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/channel/PostgresSubscribableChannel.java

```diff
@@ -110,10 +143,37 @@ protected boolean doSend(Message<?> message, long timeout) {
 
 	@Override
 	public void notifyUpdate() {
-		Message<?> message;
-		while ((message = this.jdbcChannelMessageStore.pollMessageFromGroup(this.groupId)) != null) {
-			this.dispatcher.dispatch(message);
-		}
+		this.executor.execute(() -> {
```

Comment:
Can we look into the way to schedule a task per message?
Kinda perform an execute for single poll, if it was not null, then schedule the next one.

If there is a FIFO requirements, that's where this task executor can be customized externally.

Reply:
By injecting a AsyncTaskExecutor and not an Executor we can get a CompletableFuture with the result. However getting the result (to determine whether to continue the loop) implies blocking, and is therefore equivalent to doing the loop as a single task, no matter what kind of an executor is passed in.

I can see 2 approaches (there may be more )  to achieve what you're proposing

1.  Have an AtomicReference outside the loop that is set inside the task to the result of the poll. The exit condition for the while loop is  the reference != null.  Disadvantage of this approach is that in the case of a multi-threaded executor, depending on the order of tasks finishing, we can get useless iterations through the loop if the null from a later submitted, but earlier finishing task is overridden.

2.  Ignore what is happening inside the task. The condition to exit from the while is `this.jdbcChannelMessageStore.messageGroupSize(groupId) == 0` .  There would need to be a sleep(X ms) to avoid hammering the message store. Disadvantage is that the call to notifyUpdate will take an X ms latency penalty.

Reply:
Does this not equally imply sequential message processing, irrespective of the `Executor` , since each task only submits the next one after it finishes?

Reply:
Yeah... I see now. This is indeed the same what you have just explained with a `CompletableFuture` approach.
We could swap an new task submitting with the message processing. This way we can try to pull a new message in a new thread while we process the current one.
But is it really a goal for us? We would just eliminate any FIFO from this channel logic.
It might be as an option though, or can be implemented downstream as a `bridge().channel(new ExecutorChannel())` to not care about FIFO.
But that's fully different story.

I think it is time for me to step back and agree with your current solution with a single-threaded process for the whole loop.

I'm going to pull your PR locally for the final review if you don't have anything else to add.

Thank you for your patience!

Reply:
No worries it is important to get it right (at least with our current understanding of "right"). I will address your feedback on the documentation